ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -12,137 +12,263 @@
-  private final List<OutstandingRequest> outstandingRequestList;
-  private final List<BytesValue> contactedInCurrentExecution;
+  private final FindNeighbourDispatcher findNeighbourDispatcher;
+  private Optional<RoundTimeout> currentRoundTimeout = Optional.empty();
+  private boolean iterativeSearchInProgress = false;
+  private final int maxRounds;
+
+  private int currentRound;
+
+  private final SortedMap<BytesValue, MetadataPeer> oneTrueMap = new TreeMap<>();
+
+  private final TimerUtil timerUtil;
+  private final int timeoutPeriodInSeconds;
 
   RecursivePeerRefreshState(
-      final BytesValue target,
       final PeerBlacklist peerBlacklist,
+      final Optional<NodeWhitelistController> peerWhitelist,
       final BondingAgent bondingAgent,
-      final NeighborFinder neighborFinder) {
-    this.target = target;
+      final FindNeighbourDispatcher neighborFinder,
+      final TimerUtil timerUtil,
+      final int timeoutPeriodInSeconds,
+      final int maxRounds) {
     this.peerBlacklist = peerBlacklist;
+    this.peerWhitelist = peerWhitelist;
     this.bondingAgent = bondingAgent;
-    this.neighborFinder = neighborFinder;
-    this.anteList = new ArrayList<>();
-    this.outstandingRequestList = new ArrayList<>();
-    this.contactedInCurrentExecution = new ArrayList<>();
-  }
-
-  void kickstartBootstrapPeers(final List<Peer> bootstrapPeers) {
-    for (Peer bootstrapPeer : bootstrapPeers) {
-      final BytesValue peerId = bootstrapPeer.getId();
-      outstandingRequestList.add(new OutstandingRequest(bootstrapPeer));
-      contactedInCurrentExecution.add(peerId);
-      bondingAgent.performBonding(bootstrapPeer, true);
-      neighborFinder.issueFindNodeRequest(bootstrapPeer, target);
-    }
-  }
-
-  /**
-   * This method is intended to be called periodically by the {@link PeerDiscoveryController}, which
-   * will maintain a timer for purposes of effecting expiration of requests outstanding. Requests
-   * once encountered are deemed eligible for eviction if they have not been dispatched before the
-   * next invocation of the method.
-   */
-  public void executeTimeoutEvaluation() {
-    for (int i = 0; i < outstandingRequestList.size(); i++) {
-      if (outstandingRequestList.get(i).getEvaluation()) {
-        final List<DiscoveryPeer> queryCandidates = determineFindNodeCandidates(anteList.size());
-        for (DiscoveryPeer candidate : queryCandidates) {
-          if (!contactedInCurrentExecution.contains(candidate.getId())
-              && !outstandingRequestList.contains(new OutstandingRequest(candidate))) {
-            outstandingRequestList.remove(i);
-            executeFindNodeRequest(candidate);
-          }
-        }
+    this.findNeighbourDispatcher = neighborFinder;
+    this.timerUtil = timerUtil;
+    this.timeoutPeriodInSeconds = timeoutPeriodInSeconds;
+    this.maxRounds = maxRounds;
+  }
+
+  synchronized void start(final List<DiscoveryPeer> initialPeers, final BytesValue target) {
```

Comment:
I started to add notes on some thread-safety issues I noticed, but I looked into it more and this should all actually be single-threaded within the vertx event loop.  So you should be able to cut all of the synchronized keywords.